scico/flax/train/clu_utils.py

```diff
@@ -0,0 +1,237 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2022 by SCICO Developers
+# All rights reserved. BSD 3-clause License.
+# This file is part of the SCICO package. Details of the copyright and
+# user license can be found in the 'LICENSE' file distributed with the
+# package.
+
+"""Utilities for displaying Flax models."""
+
+# These utilities have been copied from the Common Loop Utils (CLU)
+# https://github.com/google/CommonLoopUtils/tree/main/clu
+# and have been modified to remove TensorFlow dependencies
+
+# CLU is licensed under the Apache License, Version 2.0 (the "License");
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+
+import warnings
+
+warnings.simplefilter(action="ignore", category=FutureWarning)
+
+import dataclasses
+from typing import Any, Callable, Dict, List, Mapping, Optional, Sequence, Tuple, Union
+
+import numpy as np
+
+import jax
+
+import flax
+
+PyTree = Any
+ParamsContainer = Union[Dict[str, np.ndarray], Mapping[str, Mapping[str, Any]]]
+
+
+@dataclasses.dataclass
+class ParamRow:
+    """Definition of the structure of a row for printing parameters without stats."""
+
+    name: str
+    shape: Tuple[int]
+    size: int
+
+
+@dataclasses.dataclass
+class ParamRowWithStats(ParamRow):
+    """Definition of the structure of a row for printing parameters with stats."""
+
+    mean: float
+    std: float
+
+
+def flatten_dict(
+    input_dict: Dict[str, Any], prefix: str = "", delimiter: str = "/"
+) -> Dict[str, Any]:
+    """Flattens the keys of a nested dictionary.
```

Comment:
First-line docstring should be imperative, i.e. `Flatten the keys ...`. Same applies to some other functions here.

Reply:
Fixed.